backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Validator.java

```diff
@@ -1,20 +1,2310 @@
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPathDatasetDiff;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonEnvironmentBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonEnvironmentDiff;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonRequirementEnvironmentBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonRequirementEnvironmentDiff;
 import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetComponentBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetComponentDiff;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetDiff;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenVersionEnvironmentBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenVersionEnvironmentDiff;
+import ai.verta.modeldb.versioning.blob.container.BlobContainerBase;
+import ai.verta.modeldb.versioning.blob.container.ContinuousHyperparameterSetConfigContainer;
+import ai.verta.modeldb.versioning.blob.container.DiscreteHyperparameterSetConfigContainer;
+import ai.verta.modeldb.versioning.blob.container.GitCodeContainer;
+import ai.verta.modeldb.versioning.blob.container.HyperparameterSetConfigContainer;
+import ai.verta.modeldb.versioning.blob.container.PathDatasetComponentContainer;
 import io.grpc.Status;
+import java.util.List;
 
 public class Validator extends Visitor {
+
+  public Validator(AutogenBlobDiff autogenBlobDiff) throws ModelDBException {
+    if (autogenBlobDiff.getLocation().isEmpty()) {
+      throw new ModelDBException("Blob diff location is empty", Status.Code.INVALID_ARGUMENT);
+    }
+    autogenBlobDiff.preVisitDeep(this);
+    checkStatus(autogenBlobDiff.getStatus());
+  }
+
+  private void checkStatus(AutogenDiffStatusEnumDiffStatus status) throws ModelDBException {
+    checkStatus(status, () -> {}, () -> {}, () -> {});
+  }
+
+  private void checkStatus(
+      AutogenDiffStatusEnumDiffStatus diffStatus,
+      BlobContainerBase deleted,
+      BlobContainerBase added,
+      BlobContainerBase modified)
+      throws ModelDBException {
+    switch (diffStatus.toProto()) {
+      case DELETED:
+        deleted.validate();
+        break;
+      case ADDED:
+        added.validate();
+        break;
+      case MODIFIED:
+        modified.validate();
+        break;
+      default:
+        throw new ModelDBException("Unknown type", Status.Code.INVALID_ARGUMENT);
+    }
+  }
+
+  @Override
```

Comment:
You don't need to override if you're just going to use the default behavior. It also makes things much more difficult to review. Also, you don't have to implement the "deep" version of the methods unless you want to. They will call their equivalent non-deep and then call for each subfield, which is usually the desired behavior.

Reply:
fixed